### ISSUE TITLE
New Mongoid doesn't accept dots in fields

### ISSUE THREAD

**Issue by jnummelin** [kontena/kontena#2340]
The mongoid update broke RPC related pubsub as e.g. labels are using dots in the field names.
```
E, [2017-05-18T10:27:53.338743 #16] ERROR -- : Actor crashed!
api_1      | BSON::String::IllegalKey: 'io.kontena.service.id' is an illegal key in MongoDB. Keys may not start with '$' or contain a '.'.
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/string.rb:69:in `to_bson_key'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:45:in `block in to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `each'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:46:in `block in to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `each'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:49:in `block in to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `each'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `each_with_index'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:49:in `block in to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `each'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `each_with_index'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:46:in `block in to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `each'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:46:in `block in to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `each'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:49:in `block in to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `each'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `each_with_index'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:46:in `block in to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `each'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `to_bson'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/serializers.rb:163:in `serialize'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:213:in `block in serialize_fields'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:201:in `each'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:201:in `serialize_fields'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:104:in `serialize'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:193:in `block in write'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:192:in `each'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:192:in `write'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:162:in `deliver'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:108:in `block in dispatch'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/monitoring/publishable.rb:47:in `publish_command'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:107:in `dispatch'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/command/writable.rb:39:in `block in execute'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection_pool.rb:107:in `with_connection'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/server.rb:242:in `with_connection'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/command/writable.rb:38:in `execute'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/insert.rb:55:in `execute_write_command'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/write_command_enabled.rb:41:in `execute'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/collection.rb:365:in `block in insert_one'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/retryable.rb:112:in `write_with_retry'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongo-2.4.1/lib/mongo/collection.rb:356:in `insert_one'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/mongoid-5.2.1/lib/mongoid/query_cache.rb:182:in `insert_one_with_clear_cache'
api_1      |    /app/app/services/mongo_pubsub.rb:98:in `publish'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
api_1      |    /usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
````
Tracked this down to `rcp_client` channel pubsub messages:
```
api_1      | Channel: rpc_client
api_1      | data: {:type=>"request", :id=>"BZYD:ODII:NAHP:ZYZJ:FVDJ:YM2L:VKS6:EV2R:MTSM:JLH3:ZSRH:ITQ7", :message=>[0, 1322185281, "/load_balancers/remove_service", [{:id=>"591d726fd14e590010976cb6/1", :desired_state=>"running", :service_id=>"591d726fd14e590010976cb6", :service_name=>"simple-nginx", :service_revision=>1, :updated_at=>"2017-05-18 10:07:43 UTC", :instance_number=>1, :image_name=>"nginx:1-alpine", :image_credentials=>nil, :deploy_rev=>"2017-05-18 10:08:55 UTC", :stateful=>false, :user=>nil, :cmd=>nil, :entrypoint=>nil, :memory=>nil, :memory_swap=>nil, :cpu_shares=>nil, :privileged=>nil, :cap_add=>[], :cap_drop=>[], :devices=>[], :ports=>[], :volumes=>[], :volumes_from=>[], :net=>"bridge", :hostname=>"simple-nginx-1", :domainname=>"e2e.kontena.local", :exposed=>false, :log_driver=>nil, :log_opts=>{}, :pid=>nil, :wait_for_port=>nil, :env=>["KONTENA_SERVICE_ID=591d726fd14e590010976cb6", "KONTENA_SERVICE_NAME=simple-nginx", "KONTENA_GRID_NAME=e2e", "KONTENA_STACK_NAME=null", "KONTENA_NODE_NAME=moby", "KONTENA_SERVICE_INSTANCE_NUMBER=1"], :secrets=>[], :labels=>{"io.kontena.service.id"=>"591d726fd14e590010976cb6", "io.kontena.service.name"=>"simple-nginx", "io.kontena.stack.name"=>"null", "io.kontena.grid.name"=>"e2e", "io.kontena.load_balancer.name"=>"simple-lb", "io.kontena.load_balancer.internal_port"=>"80", "io.kontena.load_balancer.mode"=>"http"}, :hooks=>[], :networks=>[{:name=>"kontena", :subnet=>"10.81.0.0/16", :multicast=>true, :internal=>false}]}]]}
```

**Comment by jnummelin** [comment]
The only case this seems to happen is the `GridServices::Delete` mutation. It calls RPC `/load_balancers/remove_service` with full service_pod hash.

Although on agent side only the service name is really used from the pod:
```
service_pod = Kontena::Models::ServicePod.new(service_pod)
Celluloid::Actor[:lb_configurer].remove_config(service_pod.lb_name)
```

**Comment by SpComb** [comment]
RPC requests/notifications sent by the agent are handled directly by the `RpcServer` actor in the same worker, so the pubsub restrictions don't apply to those.

But RPC requests, notifications sent by the server, as well as responses sent by the agent, all pass through `MongoPubsub`, so the `GridServices::Delete` is not the only case.
h

#### `kontena container ls`
```
E, [2017-05-19T10:08:09.369251 #23106] ERROR -- : Actor crashed!
BSON::String::IllegalKey: 'io.kontena.container.deploy_rev' is an illegal key in MongoDB. Keys may not start with '$' or contain a '.'.
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/string.rb:69:in `to_bson_key'
        ...
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/collection.rb:356:in `insert_one'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongoid-5.2.1/lib/mongoid/query_cache.rb:182:in `insert_one_with_clear_cache'
	/home/kontena/kontena/kontena/server/app/services/mongo_pubsub.rb:94:in `publish'
```

**Comment by SpComb** [comment]
This happens because of the new [`BSON::Config.validating_keys`](http://www.rubydoc.info/github/mongodb/bson-ruby/BSON%2FConfig%3Avalidating_keys%3D) introduced in [ruby-bson version 4.1.0](https://github.com/mongodb/bson-ruby/releases/tag/v4.1.0). It's seems to default to `nil`, but then the `mongo-driver-ruby` seems to somehow set it to `true`.

There seems to be some support in the mongo driver for setting it as an option, but I can't figure out how to make that happen, just `BSON::Config.validating_keys = false` + `self.collection.insert_one({...}, validating_keys: false)` doesn't work.

The code is hard to follow: it either [defaults from `BSON::Config.validating_keys?`](https://github.com/mongodb/mongo-ruby-driver/blob/v2.4.1/lib/mongo/protocol/serializers.rb#L161), or it [defaults to false](https://github.com/mongodb/mongo-ruby-driver/blob/v2.4.1/lib/mongo/protocol/message.rb#L301), or [defaults to true](https://github.com/mongodb/mongo-ruby-driver/blob/v2.4.1/lib/mongo/protocol/insert.rb#L85) for inserts, or it's [forced to true](https://github.com/mongodb/mongo-ruby-driver/blob/v2.4.1/lib/mongo/operation/write/command/insert.rb#L58) for inserts ¯\_(ツ)_/¯

Either way, storing invalid BSON keys in BSON hashes should probably be avoided? Not sure how it has worked before, when the old mongo driver / bson gem did not validate them... was it only broken when using those fields in mongodb indexes? Perhaps we should just `MessagePack` all of the pubsub payloads as raw bytes...

**Comment by jnummelin** [comment]
Yes, it seems that they were broken only when used in indexes.

And yes, the best option seems to be some form on serialization for the pubsub payload.

**Comment by SpComb** [comment]
The following monkey-patch makes it go away:

```ruby
class Mongo::Protocol::Message
  def validating_keys?
    return false
  end
end
```

The call stack looks something like this:

```
validating_keys? options={:validating_keys=>true, :limit=>-1} @
	/home/kontena/kontena/kontena/server/app/initializers/00_mongoid.rb:16:in `validating_keys?'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:213:in `block in serialize_fields'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:201:in `each'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:201:in `serialize_fields'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:104:in `serialize'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:193:in `block in write'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:192:in `each'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:192:in `write'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:162:in `deliver'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:108:in `block in dispatch'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/monitoring/publishable.rb:47:in `publish_command'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:107:in `dispatch'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/command/writable.rb:39:in `block in execute'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection_pool.rb:107:in `with_connection'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server.rb:242:in `with_connection'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/command/writable.rb:38:in `execute'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/insert.rb:55:in `execute_write_command'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/write_command_enabled.rb:41:in `execute'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/collection.rb:365:in `block in insert_one'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/retryable.rb:112:in `write_with_retry'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/collection.rb:356:in `insert_one'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongoid-5.2.1/lib/mongoid/query_cache.rb:182:in `insert_one_with_clear_cache'
	/home/kontena/kontena/kontena/server/app/services/mongo_pubsub.rb:94:in `publish'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
```

Might be possible to override `validating_keys: false` somehow to avoid this, but maybe it's the wrong solution.
